### PR TITLE
bsunanda:Run2-alca5 Extension of bsunanda:Run2-alca4 to have updated AlCa code for IsoTrig, DiJet, GammaJet, Muon calibration for HCAL

### PR DESCRIPTION
**Pull request by bsunanda**
It is useful for IsoTrack trigger and is needed to have this for the trigger to be incorporated for Run2 HLT menu

### CODE REVIEW

**Comment by slava77** [comment]
-1

merge conflicts shouuld be resolved first

**Comment by bsunanda** [comment]
This PR was created using CMSSW_7_5_X_2015-03-05-0200 and after merge-topic with bsunanda:Run2-alca4. It did not have merge conflict when PR was requested. I wonder against which PR it has conflict. Could you let me know against which branch should I test the PR?
________________________________

**Comment by slava77** [comment]
Hi Sunanda

You can merge CMSSW_7_5_X first in a test release and then cms-merge-topic
your branch.  It should tell where the problem is.

**Comment by bsunanda** [comment]
I created to release areas: CMSSW_7_5_DEVEL_X_2015-03-07-0200 and CMSSW_7_5_X_2015-03-07-0200. I tried cms-merge-topic 8131 to either of these and the merge was complete and the codes compiled fine. There are 3 .h files which were redundant and removed. Please help me with which version I should try to merge.   Sunanda
________________________________

**Comment by davidlange6** [comment]
likely it is a conflict with your other pR that got merged yesterday. I would suggest to try again on -03-08-0200 which is now available.

**Comment by davidlange6** [comment]
@bsunanda - any news on this PR?

**Comment by bsunanda** [comment]
Can I upload things with corrections? I can do that today
________________________________

**Comment by davidlange6** [comment]
Yes, just update your branch and the PR will update with your changes.

________________________________

**Comment by cvuosalo** [comment]
@cmsbuild please test

Comparisons never completed. Maybe they will complete this time.

**Comment by diguida** [comment]
@bsunanda can you please rebase?

File: DataFormats/HcalIsolatedTrack/interface/HcalIsolatedTrackCandidate.h [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    void SetEtaPhiHcal(double eta, double phi, int ieta, int iphi) {
+      etaHcal_=eta; phiHcal_=phi; ietaHcal_=ieta; iphiHcal_=iphi; etaPhiHcal_=true;
+    }
+    std::pair<double,double> EtaPhiHcal() const {
+      return ((etaPhiHcal_) ? std::pair<double,double>(etaHcal_,phiHcal_) : std::pair<double,double>(0,0));
+    }
+    std::pair<int,int> towerIndex() const {
+      return ((etaPhiHcal_) ? std::pair<int,int>(ietaHcal_,iphiHcal_) : std::pair<int,int>(0,0));
+    }
+    bool etaPhiHcal() const {return etaPhiHcal_;}
```
Comment (cvuosalo) [inline DataFormats/HcalIsolatedTrack/interface/HcalIsolatedTrackCandidate.h]: Using similar names for two methods that differ only in their capitalization is confusing and prohibited by the CMS coding [standards] (https://cms-cpt-software.web.cern.ch/cms-cpt-software/General/CodingAndStyleRules.pdf). This class has etaPhiHcal / EtaPhiHcal and etaPhiEcal / EtaPhiEcal. The boolean version of the method should be called etaPhiEcalValid() or something like that. The coding standard prohibits methods that start with a capital letter, but there are six such method names in this class (SetMaxP, SetEnergyEcal, SetEtaPhiEcal, SetEtaPhiHcal, EtaPhiHcal, EtaPhiEcal). The problems with this code were noted a month ago in a previous pull request (#7523). Now is a good time to finally fix them.

File: DataFormats/HcalIsolatedTrack/src/IsolatedPixelTrackCandidate.cc [hunk 2]
```diff
@@ -34,33 +34,31 @@ bool IsolatedPixelTrackCandidate::overlap( const Candidate & c ) const {
...
-  if (fabs(track_->eta())<1.740) 
-    {
-      for (int i=0; i<72; i++)
-	{
-	  if (delta<(i+1)*0.087266462&&delta>i*0.087266462) iphi=i;
-	}
+  int ieta(0), iphi(0), nphi(72), kphi(1);
+  double etas[24]={0.000,0.087,0.174,0.261,0.348,0.435,0.522,0.609,0.696,
+		   0.783,0.870,0.957,1.044,1.131,1.218,1.305,1.392,1.479,
+		   1.566,1.653,1.740,1.830,1.930,2.043};
```
Comment (cvuosalo) [inline DataFormats/HcalIsolatedTrack/src/IsolatedPixelTrackCandidate.cc]: Please do not duplicate these values of the calo towers boundaries. Include the file RecoJets/JetAnalyzers/interface/CaloTowerBoundries.h to get these eta values.

File: DataFormats/HcalIsolatedTrack/src/IsolatedPixelTrackCandidate.cc [hunk 3]
```diff
@@ -34,33 +34,31 @@ bool IsolatedPixelTrackCandidate::overlap( const Candidate & c ) const {
...
-	{
-	  if (delta<2*(i+1)*0.087266462&&delta>2*i*0.087266462) iphi=2*i;
-	}
+  }
+
+  double phi = track_->phi();
+  if (phi < 0) phi += (2*M_PI);
+  double delta = phi+(kphi*0.087266462);
+  for (int i=0; i<nphi; i++) {
+    if (delta<=(kphi*(i+1)*0.087266462)) {
```
Comment (cvuosalo) [inline DataFormats/HcalIsolatedTrack/src/IsolatedPixelTrackCandidate.cc]: Please use a const double in place of this magic number 0.087266...

**Comment by cvuosalo** [comment]
+1

For #8131 98dd253563c5c34ef02f2563cd5bbe1513ef1aa4

Updates of several calibration schemes.

With #8347 most of the coding issues raised above have been addressed. There is still the issue of the calo towers eta boundary values being duplicated in the code, but that can wait for a later PR. Otherwise, the code changes are satisfactory, as are the Jenkins results. Short matrix tests against baseline CMSSW_7_5_X_2015-03-12-1400 show no significant differences.

File: Calibration/HcalAlCaRecoProducers/src/AlCaIsoTracksProducer.cc [hunk 4]
```diff
@@ -1,206 +1,206 @@
...
+  selectionParameters.maxDxyPV        = iConfig.getParameter<double>("MaxDxyPV");
+  selectionParameters.maxDzPV         = iConfig.getParameter<double>("MaxDzPV");
+  selectionParameters.maxChi2         = iConfig.getParameter<double>("MaxChi2");
+  selectionParameters.maxDpOverP      = iConfig.getParameter<double>("MaxDpOverP");
+  selectionParameters.minOuterHit     = iConfig.getParameter<int>("MinOuterHit");
+  selectionParameters.minLayerCrossed = iConfig.getParameter<int>("MinLayerCrossed");
+  selectionParameters.maxInMiss       = iConfig.getParameter<int>("MaxInMiss");
+  selectionParameters.maxOutMiss      = iConfig.getParameter<int>("MaxOutMiss");
+  a_coneR                             = iConfig.getParameter<double>("ConeRadius");
+  a_charIsoR                          = a_coneR + 28.9;
```
Comment (diguida) [inline Calibration/HcalAlCaRecoProducers/src/AlCaIsoTracksProducer.cc]: As already noticed by @cvuosalo please do not use magic numbers: what does `28.9` means?

File: Calibration/HcalAlCaRecoProducers/src/AlCaIsoTracksProducer.cc [hunk 5]
```diff
@@ -208,354 +208,274 @@ AlCaIsoTracksProducer::AlCaIsoTracksProducer(const edm::ParameterSet& iConfig)
...
+  }
+  edm::Handle<HBHERecHitCollection> hbhe;
+  iEvent.getByToken(tok_hbhe_, hbhe);
+  if (!hbhe.isValid()) {
+    edm::LogWarning("HcalIsoTrack") << "Cannot access the collection " << labelHBHE_;
+    return;
+  }
+
+  //Get L1 trigger object
+  ptL1 = etaL1 = phiL1 = 0;
```
Comment (diguida) [inline Calibration/HcalAlCaRecoProducers/src/AlCaIsoTracksProducer.cc]: `ptL1, etaL1, phiL1` are data members used for producing collections and selecting on them, which are reset at every event in the `produce` function member. Anyhow, they are not initialised in the c'tor. Can you do that?

File: Calibration/HcalAlCaRecoProducers/test/AlCaDiJetsProducer_cfg.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import FWCore.ParameterSet.Config as cms
+
+process = cms.Process("MYDIJETS")
+
+process.load("Configuration.Geometry.GeometryIdeal_cff")
+process.load("Configuration.StandardSequences.MagneticField_cff")
+process.load('Configuration.StandardSequences.FrontierConditions_GlobalTag_cff')
```
Comment (diguida) [inline Calibration/HcalAlCaRecoProducers/test/AlCaDiJetsProducer_cfg.py]: please use the `condDBv2` version

File: Calibration/HcalAlCaRecoProducers/test/AlCaDiJetsProducer_cfg.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+from Configuration.AlCa.autoCond import autoCond
+process.GlobalTag.globaltag=autoCond['startup']
```
Comment (diguida) [inline Calibration/HcalAlCaRecoProducers/test/AlCaDiJetsProducer_cfg.py]: please use the `condDBv2` version
The correct key for `autoCond_condDBv2` dictionary is `run2_mc`.

**Comment by diguida** [comment]
@bsunanda please rebase, github is complaining about merge conflicts
Please do not forget to address my comments also in this PR

**Comment by diguida** [comment]
The comments are addressed in last two commits of #8347

**Comment by slava77** [comment]
there are merge conflicts here now

**Comment by bsunanda** [comment]
Hi

I rebased this sometime ago to avoid merge conflicts and it again has created a merge conflict. I am rebasing again without any further improvement. Please merge this quickly before other PR's.

Sunanda
________________________________
